### PR TITLE
[SPARK-27504][SQL] File source V2: support refreshing metadata cache

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

In file source V1, if some file is deleted manually, reading the DataFrame/Table will throws an exception with suggestion message 
```
It is possible the underlying files have been updated. You can explicitly invalidate the cache in Spark by running 'REFRESH TABLE tableName' command in SQL or by recreating the Dataset/DataFrame involved.
```
After refreshing the table/DataFrame, the reads should return correct results.

We should follow it in file source V2 as well.
## How was this patch tested?
Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
@cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala [hunk 1]
```diff
@@ -60,14 +61,14 @@ class MetadataCacheSuite extends QueryTest with SharedSQLContext {
     }
   }
 
-  test("SPARK-16337 temporary view refresh") {
+  test("SPARK-16337,SPARK-27504 temporary view refresh") {
     withTempView("view_refresh") { withTempPath { (location: File) =>
       // Create a Parquet directory
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala]: Let's update this comment if happening to edit this file in the future.

**Comment by rdblue** [comment]
@gengliangwang, this change prevents moving DataSourceV2Relation into Catalyst. I don't think this should have been merged because logical plan nodes, including relations, are immutable and should not expose methods that alter state.

**Comment by gengliangwang** [comment]
@rdblue Make sense. I will remove it.

**Pull request closed without merging**